sigstore-java/src/main/java/dev/sigstore/StringMatchable.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2024 The Sigstore Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package dev.sigstore;
+
+import com.google.common.base.Preconditions;
+import java.util.regex.Pattern;
+
+/**
+ * An interface for allowing direct string matching or regular expressions. Use the static factory
+ * {@link #of(String)} or {@link #of(Pattern)} to instantiate the appropriate matcher.
+ */
+public interface StringMatchable {
+
+  boolean matches(String target);
+
+  /** The input provided to instantiate the string matchable. */
+  String source();
```

Comment:
What is the purpose of `source()` if the user can't tell if the resulting string is a regexp or literal?

Reply:
That's a good point, I can clarify that.

Reply:
I had it as a functional interface at first, but it was hard to provide good error messaging when things didn't match. But the impl of source can make that more clear.

Reply:
`source()` can be confusing if the string contains special characters.
For instance, consider `.of(".*")`.
The source would look like `.*`, and everybody would think it is a regexp.

I would rather have `.toString()` so the value represents better in the debugger.